src/main/kotlin/com/atlassian/performance/tools/report/api/judge/RelativeNonparametricPerformanceJudge.kt

```diff
@@ -48,21 +66,21 @@ class RelativeNonparametricPerformanceJudge(
         val baselineCohort = baselineResult.cohort
         val experimentCohort = experimentResult.cohort
         val reportName = "Regression for $label $experimentCohort vs $baselineCohort"
-        val reader = ActionMetricsReader()
```

Comment:
Why does it need to be injectable?

Reply:
My understanding is that reader is not responsibility of `RelativeNonparametricPerformanceJudge` so if it relies on that reader then the reader should be it's dependency.

If we instantiate the reader as part of the judge class it technically is not different than having a private method which would handle the functionality. The class in this form is not unit testable and communication between judge and reader can't be observed nor intercepted. The solution is not modular and without known to me reason limits usability of existing in the lib functionality.

Moreover injecting dependencies to the class is honest with it's consumer. We don't end up with forced dependencies (e.g. I don't want log4j i my classpath) if consumer can replace them. Reusability is potentially higher.

Reply:
I agree with the arguments. I wonder whether it's a "just-in-case" refactor or is there a specific use case for it?
All the internal `ActionMetricsReader` forces is `filter { it.result == ActionResult.OK }`. The rest is just transforming data to other types, which is forced by the method signature (before and after the refactor).

Reply:
> Why does it need to be injectable?
Because it's not testable. To prepare a unit test you have to know the logic behind ActionMetricsReader and that becomes an integration test, not a unit test. As a result - much more time needed to make a unit test
I changed it so I could provide a unit test

Reply:
Ah, then it's not needed. You can reuse the same pattern as `IndependentCohortsJudgeTest` does: it loads real `EdibleResult`s quite easily.

Reply:
> Is it a "just-in-case" refactor or is there a specific use case for it?

Possibly it's "just-in-case", however as @mgrzaslewicz mentioned I also see how not excluding the reader's functionality forces us to test it together with the judge. 

I believe we are doing many things "just-in-case". Designing architecture or following good practices are done "just-in-case", because we did write software before, we did consume libraries before and we know painpoints of not doing things "just-in-case".

> Do we need to override the OK filtering?

Perhaps. It depends on what is the responsibility of the `RelativeNonparametricPerformanceJudge`, if it's not the knowledge of OK filtering then we should delegate it.

Sometimes you can model the responsibility very thick, however here we already decided that it's good to delegate part of `RelativeNonparametricPerformanceJudge`'s functionality when `ActionMetricsReader` was created. I think it's natural sign that it should be completely separated and proper communication layer (interface) should be esabilished.

Reply:
> here we already decided that it's good to delegate part of RelativeNonparametricPerformanceJudge's functionality when ActionMetricsReader was created

It was a simple deduplication. Sure it can also be used for injectability, but does anyone actually need to override this inlined behavior? Tests don't need that, see all the existing tests here.

Reply:
I find the 3 last paragraphs from https://www.martinfowler.com/bliki/Yagni.html in some sense related

Reply:
> Sure it can also be used for injectability as a side-effect, but does anyone actually need to override this inlined behavior? Tests don't need that, see all the existing tests here.

I find most of constructor invocations instead of injecting the object or provider a yellow light.
We don't need injectability per se. It's not just in case. It's mostly a must-have.
We need maintanability by having code easy to test and follow good programming principles - injectable behaviour is one of them

Reply:
> We need maintanability by having code easy to test and follow good programming principles - injectable behaviour is one of them

I strongly agree. To be precise I see injectability as implementation of Open-Closed principle.